jxl/src/image/internal.rs

```diff
@@ -0,0 +1,283 @@
+    pub(super) unsafe fn row_mut(&mut self, row: usize) -> &mut [MaybeUninit<u8>] {
+        assert!(row < self.num_rows);
+        let start = row * self.bytes_between_rows;
+        // SAFETY: `start` is guaranteed to be <= isize::MAX, and `self.buf + start` is guaranteed
+        // to fit within the same allocated object, as per safety invariants of this struct.
+        // We checked above that `row` and `cols` satisfy the requirements to apply the safety
+        // invariant.
+        let start = unsafe { self.buf.add(start) };
+        // SAFETY: due to the struct safety invariant, we know the entire slice is in a range of
+        // memory valid for writes. Moreover, the caller promises not to write uninitialized data
+        // in the returned slice. Finally, the caller guarantees aliasing rules will not be violated.
+        unsafe { std::slice::from_raw_parts_mut(start, self.bytes_per_row) }
+    }
+
+    /// # Safety
+    /// The caller must ensure that ownership and lifetime rules are respected (for example,
+    /// because they have shared access to the data).
+    pub(super) unsafe fn row(&self, row: usize) -> &[MaybeUninit<u8>] {
+        assert!(row < self.num_rows);
+        let start = row * self.bytes_between_rows;
+        // SAFETY: `start` is guaranteed to be <= isize::MAX, and `self.buf + start` is guaranteed
+        // to fit within the same allocated object, as per safety invariants of this struct.
+        // We checked above that `row` and `cols` satisfy the requirements to apply the safety
+        // invariant.
+        let start = unsafe { self.buf.add(start) };
+        // SAFETY: due to the struct safety invariant, we know the entire slice is in a range of
+        // memory valid for writes. Moreover, the caller promises not to write uninitialized data
+        // in the returned slice. Finally, the caller guarantees aliasing rules will not be violated.
+        unsafe { std::slice::from_raw_parts(start, self.bytes_per_row) }
+    }
+
+    /// Extracts a sub-rectangle from this buffer. Rectangle coordinates are in bytes.
+    /// Safety note: the returned RawImageBuffer retains the same kind of access (read or write) as `self`.
+    pub(super) fn rect(&self, rect: Rect) -> Self {
+        if rect.size.0 == 0 || rect.size.1 == 0 {
+            return Self::empty();
+        }
+        assert!(rect.origin.1 < self.num_rows);
+        assert!(rect.origin.1.checked_add(rect.size.1).unwrap() <= self.num_rows);
+        assert!(rect.origin.0 < self.bytes_per_row);
+        assert!(rect.origin.0.checked_add(rect.size.0).unwrap() <= self.bytes_per_row);
+        // SAFETY: the safety invariant of `self`, together with the above check, guarantees that
+        // the calculation does not overflow and that the new pointer stays within the bounds of
+        // the allocation.
+        let start_ptr = unsafe {
+            self.buf
+                .add(rect.origin.1 * self.bytes_between_rows + rect.origin.0)
```

Comment:
Is it expected that `...origin.1` is expressed in pixels, but `...origin.0` is expressed in bytes?  Should `...origin.0` be multiplied by `bytes_between_pixels` or `bytes_per_pixel`?